File: healthcare-app-v3/services/aof-service/app/nats_client.py
from nats.aio.client import Client as NATS
from app.logger import logger
import os
import asyncio

class NATSClient:
    def __init__(self):
        self.nc = NATS()

    async def connect(self):
        """Connect to the NATS server."""
        server = os.getenv("NATS_SERVER", "nats://localhost:4222")

        async def disconnected_cb():
            logger.warning("Disconnected from NATS server.")

        async def reconnected_cb():
            logger.info("Reconnected to NATS server.")

        async def error_cb(e):
            logger.error(f"Error occurred: {e}")

        async def closed_cb():
            logger.error("Connection to NATS server is closed.")

        retries = 0
        while True:
            try:
                await self.nc.connect(
                    servers=[server],
                    reconnected_cb=reconnected_cb,
                    disconnected_cb=disconnected_cb,
                    error_cb=error_cb,
                    closed_cb=closed_cb,
                    max_reconnect_attempts=-1,  # Infinite reconnect attempts
                )
                logger.info(f"Connected to NATS server at {server}.")
                return
            except Exception as e:
                retries += 1
                logger.warning(f"Attempt {retries}: Failed to connect to NATS. Retrying in 5 seconds... Error: {e}")
                await asyncio.sleep(5)

File: healthcare-app-v3/services/transcribe-service/src/messaging/nats-client.ts
import { connect, NatsConnection } from "nats";
import logger from "../utils/logger";

export async function initializeNatsClient(): Promise<NatsConnection> {
  const natsUrl = process.env.NATS_SERVER || "nats://localhost:4222";
  const maxRetries = 5;
  const retryDelay = 5000; // 5 seconds

  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      const natsClient = await connect({ servers: natsUrl });
      logger.info("Connected to NATS server:", natsUrl);
      return natsClient;
    } catch (error) {
      logger.warn(
        `Failed to connect to NATS server. Attempt ${attempt} of ${maxRetries}. Retrying in ${retryDelay}ms...`
      );
      if (attempt === maxRetries) {
        logger.error("Unable to connect to NATS after multiple attempts.");
        throw error;
      }
      await new Promise((resolve) => setTimeout(resolve, retryDelay));
    }
  }

  throw new Error("Unable to connect to NATS after all retry attempts.");
}